_posts/blog/2019-09-17-erasure-coding.md

```diff
@@ -0,0 +1,178 @@
+an EC directory were as much as 3 times faster than a replication directory 
+, and reads were up to 2 times faster.
+
+Another side effect of EC is loss of data locality.  For performance reasons, EC
+data blocks are striped, so multiple DataNodes must be contacted to read a single
+block of data.  For large sequential reads this is not a
+problem, but it can be an issue for small random lookups.  For the latter case,
+using RS 6-3 with 64KB stripes mitigates some of the random lookup pain
+without compromising sequential read/write performance.
+
+#### Important Warning
+
+Before continuing, an important caveat;  the current implementation of EC on Hadoop supports neither hsync
+nor hflush.  Both of these operations are silent no-ops (EC [limitations]).  We discovered this the hard
+way when a data center power loss resulted in write-ahead log corruption, which were
+stored in an EC directory.  To avoid this problem ensure all 
+WAL directories use replication.  It's probably a good idea to keep the
+accumulo namespace replicated as well, but we have no evidence to back up that assertion.  As with all
+things, don't test on production data.
+
+### EC Performance
+
+To test EC performance, we created a series of clusters on AWS.  Our Accumulo stack consisted of
+Hadoop 3.1.1 built with the Intel ISA-L library enabled, Zookeeper 3.4.13, and Accumulo 1.9.3 configured
+to work with Hadoop 3 (we did our testing before the official release of Accumulo 2.0). The encoding
+policy is set per-directory using the [hdfs] command-line tool. To set the encoding policy
+for an Accumulo table, first find the table ID (for instance using the Accumulo shell's
+"table -l" command), and then from the command line set the policy for the corresponding directory
+under /accumulo/tables.  Note that changing the policy on a directory will set the policy for
+child directories, but will not change any files contained within.  To change the policy on an existing
+Accumulo table, you must first set the encoding policy, and then run a major compaction to rewrite
+the RFiles for the table.
+
+Our first tests were of sequential read and write performance straight to HDFS.  For this test we had
+a cluster of 32 HDFS nodes (c5.4xlarge [AWS] instances), 16 Spark nodes (r5.4xlarge),
+3 zookeepers (r5.xlarge), and 1 master (r5.2xlarge).
+
+The first table below shows the results for writing a 1TB file.  The results are the average of three runs
+for each of the directory encodings Reed-Solomon (RS) 6-3 with 64KB stripes, RS 6-3 with 1MB stripes,
+RS 10-4 with 1MB stripes, and the default triple replication.  We also varied the number of concurrent
+Spark executors, performing tests with 16 executors that did not stress the cluster in any area, and with
+128 executors which exhausted our network bandwidth allotment of 5 Gbps. As can be seen, in the 16 executor
+environment, we saw greater than a 3X bump in bandwidth using RS 10-4 with 1MB stripes over triple replication.
```

Comment:
I find this wording slightly confusing.  If  I understand correctly EC should use less bandwidth across the entire cluster, however the phrase `we saw greater than a 3X bump in bandwidth using RS 10-4` reads like EC used more bandwidth.

Reply:
guess I should change bandwidth to throughput.  You're correct that there should be less aggregate bandwidth used throughout the cluster, since less data is being written overall.  How that traffic is distributed throughout the cluster varies greatly between replication and EC.  With EC, the vast majority of traffic is from client nodes to HDFS datanodes, with very little traffic between HDFS nodes.  With replication, the client->datanode traffic falls off, but the datanode->datanode traffic is double that.  The AWS monitoring tools show that quite well...maybe it's worth including some screen grabs I did during testing?

Reply:
changed bandwidth...I guess the rest is getting into the weeds again.